oauth2l-go/src/oauth2client/oauth2client.go

```diff
@@ -0,0 +1,313 @@
+	return s.pkey.Sign(rand, msg, opts)
+}
+
+// Default function to handle authorize flow. This hanlder prints the URL on
+// console and reads verification from stdin (i.e. let user paste verification
+// code to console).
+func defaultAuthorizeFlowHandler(authorizeUrl string) string {
+	// Print the url on console, let user authorize and paste the token back.
+	fmt.Printf("Go to the following link in your browser:\n\n   %s\n\n",
+		authorizeUrl)
+
+	fmt.Println("Enter verification code: ")
+	code := ""
+	fmt.Scanln(&code)
+	return code
+}
+
+// Run the three-legged oauth authorize flow.
+func authorizeFlow(secret map[string]interface{}, scopes string, handler func(string) string) string {
+	// Marshaw a url to be printed on console. In web based oauth flow, the
+	// browser should redirect the user to this url
+	paramsMap := map[string]string{
+		"access_type":                 "offline",
+		"auth_provider_x509_cert_url": "",
+		"redirect_uri":                oobCallbackUrn,
+		"response_type":               "code",
+		"client_id":                   "",
+		"scope":                       scopes,
+		"project_id":                  "",
+	}
+
+	params := url.Values{}
+	for key := range paramsMap {
+		if val, ok := secret[key]; ok {
+			params.Add(key, val.(string))
+		} else {
+			params.Add(key, paramsMap[key])
+		}
+	}
+
+	authorizeUrl := secret["auth_uri"].(string) + "?" + params.Encode()
+
+	return handler(authorizeUrl)
+}
+
+// Run the three-legged oauth verification. Sends a request to auth_uri
+// containing the verification code.
+func verifyFlow(secret map[string]interface{}, scopes string, code string) (*Token, error) {
+	// Construct a POST request to fetch oauth token using the verificaton code.
+	postData := map[string]string{
+		"client_id":    secret["client_id"].(string),
+		"code":         code,
+		"scope":        scopes,
+		"grant_type":   "authorization_code",
+		"redirect_uri": oobCallbackUrn,
+	}
+	if clientSecret, ok := secret["client_secret"]; ok {
+		postData["client_secret"] = clientSecret.(string)
+	}
+
+	params := url.Values{}
+	for key := range postData {
+		params.Add(key, postData[key])
+	}
+
+	// Send the POST request.
+	response, err := http.PostForm(secret["token_uri"].(string), params)
+	if err != nil {
+		return nil, fmt.Errorf("Failed in http request to verify verification code.")
+	}
+	defer response.Body.Close()
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read http body for verification request.")
+	}
+	var token *Token
+	if err := json.Unmarshal(body, &token); err != nil {
+		return nil, fmt.Errorf("Failed to parse json in verify flow.")
+	}
+
+	return token, nil
+}
+
+// Helper struct used in sign JWT
+type signerOpts struct{}
+
+func (r signerOpts) HashFunc() crypto.Hash {
+	return crypto.SHA256
+}
+
+// Base 64 encode a block. The output doesn't contain the trailing double equal
+// signs.
```

Comment:
The https://g3doc.corp.google.com/go/g3doc/go-readability-review-comments.md?cl=head didn't seem to mention line wrap for comments. let's keep 80 chars for now for consistency.